snuba/reader.py

```diff
@@ -0,0 +1,98 @@
+from __future__ import annotations
+
+import re
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import TYPE_CHECKING, Any, Mapping, Optional, Sequence
+
+from dateutil.tz import tz
+
+
+if TYPE_CHECKING:
+    from mypy_extensions import TypedDict
+
+    Column = TypedDict("Column", {"name": str, "type": str})
+    Result = TypedDict(
+        "Result",
+        {
+            "meta": Sequence[Column],
+            "data": Sequence[Mapping[str, Any]],
+            "totals": Mapping[str, Any],
+        },
+        total=False,
+    )
+
+
+class Reader(ABC):
+    @abstractmethod
+    def execute(
+        self,
+        query: str,
+        settings: Optional[Mapping[str, str]] = None,
+        query_id: Optional[str] = None,
+        with_totals: bool = False,
+    ) -> Result:
+        """Execute a query."""
+        raise NotImplementedError
+
+
+class NativeDriverReader(Reader):
+
+    DATE_TYPE_RE = re.compile(r"(Nullable\()?Date\b")
+    DATETIME_TYPE_RE = re.compile(r"(Nullable\()?DateTime\b")
+
+    def __init__(self, client):
+        self.__client = client
+
+    def __transform_result(self, result, with_totals: bool) -> Result:
```

Comment:
This is mostly a direct copy of `scrub_ch_data`, with the exception that the `with_totals` handling at the end is brought over from `raw_query`.

Reply:
I guess it'd be `Tuple[Sequence[Tuple[Any]], Sequence[Tuple[str, str]]]`, per https://github.com/mymarilyn/clickhouse-driver/blob/464d232a7e4caf75608e6507175e0bc6ac813f12/clickhouse_driver/client.py#L185-L189

Reply:
Did you leave it out intentionally because it comes from non typed code anyway ?

Reply:
Yeah, I figured this was probably not super useful right now, though if/when we start using mypy with CI it might be more beneficial (if this code is still around.)